### PR TITLE
Fix intellisense errors by moving TerminalApp projects around

### LINKED ISSUE #6866: Extra IntelliSense errors caused by include paths and precompiled headers
# Environment

```none
Windows build number: 10.0.18363.0
Windows Terminal version (if applicable): master branch, commit fc083296b98a2c45126da7a0a61a2ccb37786e48
Visual Studio 2019, 16.6
```

# Steps to reproduce

* Clone the project as described in the README.
* Open `OpenConsole.sln` in Visual Studio, switch the platform from "Any CPU" to "x64," and build the TerminalAppLib project.
* Open `TerminalPage.cpp` (under `controls/TerminalPage.xaml` in the solution explorer).

# Expected behavior

IntelliSense should not show any errors/red squiggles, given that the build succeeded.

# Actual behavior

IntelliSense finds the wrong `pch.h` file and goes on to produce a few hundred errors about missing definitions.

Specifically, it finds `src\cascadia\TerminalApp\pch.h` instead of `src\cascadia\TerminalApp\lib\pch.h`. You can confirm this by using Go To Definition/F12 on the `#include "pch.h"` at the top of `TerminalPage.cpp`. Changing this to `#include "lib/pch.h"`, or making an equivalent change to the project include paths, lets IntelliSense find the right header and all the errors disappear.

The build succeeds because the compiler does not use the usual include path search for this header when PCH is enabled, while IntelliSense fails because its PCH implementation does not work this way.

### PR DESCRIPTION
**Pull request by zadjii-msft**
The easiest fix was actually just moving all the source files from
`TerminalApp` to `TerminalApp/lib`, where the appropriate `pch.h`
actually resides.

Closes #6866

### CODE REVIEW

**Comment by rpjohnst** [comment]
FYI in case you are interested in pushing this further- this approach will prevent IntelliSense from actually pre-compiling that header, losing out on that optimization. This means longer pauses after first opening a file, and after making "top-level" edits (i.e. outside the body of a function).

The rule of thumb we (the IntelliSense team) use here is that the build should still work with PCH disabled, and then IntelliSense should automatically work as well. If I understand correctly, then adding `lib` to TerminalAppLib's include paths shouldn't cause any conflicts (as there are no other header files there), and would allow everything to work.

**Comment by miniksa** [comment]
> FYI in case you are interested in pushing this further- this approach will prevent IntelliSense from actually pre-compiling that header, losing out on that optimization. This means longer pauses after first opening a file, and after making "top-level" edits (i.e. outside the body of a function).
> 
> The rule of thumb we (the IntelliSense team) use here is that the build should still work with PCH disabled, and then IntelliSense should automatically work as well. If I understand correctly, then adding `lib` to TerminalAppLib's include paths shouldn't cause any conflicts (as there are no other header files there), and would allow everything to work.

Neither @zadjii-msft or I can tell whether your suggestion means "use the include path INSTEAD of adding the extra PCH to all of the headers" or "use the include path IN ADDITION TO adding the extra PCH to all of the headers." Can you please clarify, @rpjohnst?

**Comment by zadjii-msft** [comment]
And you know what, I did try just adding `$(OpenConsoleDir)\cascadia\TerminalApp\lib\;` to `<AdditionalIncludeDirectories>` in `TerminalApp.vcxproj`, and removing the `lib/pch.h` line from TerminalPage.cpp, but that gave me the same list of Intellisense errors as always. Maybe I'm doing it wrong?

**Comment by rpjohnst** [comment]
Ah, sorry for the confusion. I meant to suggest using the include path *instead* of adding the extra `#include` lines.

And now that you mention that you've tried that, I think I see why it didn't work- `#include "pch.h"` with "" instead of <> prioritizes the source file's directory over the include search paths, and the wrong PCH happens to be in the source file's directory.

So I tried adding `$(OpenConsoleDir)\cascadia\TerminalApp\lib` *and* replacing `#include "pch.h"` with `#include <pch.h>"` and that seems to be working. (Well, I'm now hitting a different bug where IntelliSense crashes on some coroutines stuff, which `TerminalPage.cpp` is full of, but at least it's not spitting out errors 🙂)

**Comment by zadjii-msft** [comment]
Okay, weird. I've tried that too, but now it seems that whenever I open a cpp file with the `<pch.h>` in it, it will intermittently have intellisense errors in it, until I load up the header for that file. Once intellisense has had a chance to parse the header too, it seems like the errors in the cpp file go away as well. 

I have no idea if that's expected or not.

**Comment by rpjohnst** [comment]
Hm, that's definitely not expected. A guess at what's happening- perhaps the cpp file you open finds the wrong `pch.h`, while the header finds the correct one. IntelliSense could then build a working PCH in response to opening the header, which could then be picked up by the cpp file when you return to it.

If so, one thing to double check is the order of the `AdditionalIncludeDirectories` list. Make sure `lib` is before the existing `..`. (Or even replace it, if that doesn't break anything else?) It might also help to temporarily disable PCH use project-wide and make sure the compiler can still find everything.

If that doesn't help, then maybe it's just some leftover state that would go away if you did "Project > Rescan File," restarted VS, and/or deleted the .vs directory. (Fair warning, that will reset a few things like the set of open files; if that's a problem then it may be enough to delete just `.vs/<Project>/v16/ipch`)

**Comment by miniksa** [comment]
K well. I was having Intellisense issues and I merged this into my branch and they're gone. Well, Intellisense is still slow but that's par for a WinRT project. So I don't care if there's a better way, this is better than 100% broken as it is now.

**Comment by DHowett** [comment]
It might be even easier to move the _lib_ project into the `TerminalApp/` folder and make a `dll` project that just dllizes it

File: src/cascadia/TerminalApp/dll/TerminalApp.vcxproj [hunk 1]
```diff
@@ -28,15 +28,15 @@
...
-    <ClInclude Include="TabRowControl.h" />
-    <ClInclude Include="App.h" />
-    <ClInclude Include="Tab.h" />
+    <ClInclude Include="../TerminalPage.h" />
+    <ClInclude Include="../MinMaxCloseControl.h" />
+    <ClInclude Include="../AppKeyBindings.h" />
+    <ClInclude Include="../TitlebarControl.h" />
+    <ClInclude Include="../TabRowControl.h" />
+    <ClInclude Include="../App.h" />
+    <ClInclude Include="../Tab.h" />
```
Comment (DHowett) [inline src/cascadia/TerminalApp/dll/TerminalApp.vcxproj:39]: does the dll project even need the ClIncludes? Can it live w/o them?

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/dll/TerminalApp.vcxproj:39]: probably?

**Pull request merged**